lib/custom-types.js

```diff
@@ -0,0 +1,136 @@
+'use strict';
+var RSVP = require('rsvp');
+var _ = require('lodash');
+
+exports.mapCustomTypes = function(schema, types){
+
+  return _.reduce(Object.keys(schema), mapCustomTypes(schema), []);
+
+  function isEmbeddedArray(schema, key){
+    return _.isArray(schema[key]) && //And IS an array
+      _.isObject(schema[key][0]) && //With schema placed inside
+      ! (
+        schema[key][0].ref || //Which is not a reference configuration
+        schema[key][0].type //Neither it is array of Strings or Dates for instance
+      )
+  }
+
+  function isEmbeddedSchema(schema, key){
+    return  _.isPlainObject(schema[key]) && //And is not null or some other typeof quirk
+      !_.isArray(schema[key]) && //And not an array of embedded docs
+        !(
+          schema[key].ref || //And neither reference
+          schema[key].type // nor end path configuration
+        );
+  }
+
+
+  function mapCustomTypes(schema, prefix){
+    return function(memo, key){
+      if((typeof schema[key]).toLowerCase() === "string") {
+        var customType = types[schema[key]];
+        if(!customType) throw new Error("Custom Type \"" + schema[key] + "\" is not defined");
+        memo.push({
+          path: prefix ? [prefix, key].join('.') : key,
+          hooks: customType.hooks,
+          schema: customType.schema,
+          typeId: schema[key],
+          type: customType
+        });
+      } else if (isEmbeddedArray(schema, key)) {
+        memo = memo.concat(_.reduce(
+          Object.keys(schema[key][0]),
+          mapCustomTypes(
+            schema[key][0],
+            prefix ?
+              [prefix, key + '.0'].join('.') :
+              key + '.0'
+          ),
+          []
+        ));
+      } else if (isEmbeddedSchema(schema, key)) {
+        memo = memo.concat(_.reduce(
+          Object.keys(schema[key]),
+          mapCustomTypes(
+            schema[key],
+            prefix ?
+              [prefix, key].join('.') :
+              key
+          ),
+          []
+        ));
+      }
+      return memo;
+    }
+  }
+};
+
+exports.rewriteSchema = function(schema, paths){
+  paths.forEach(function(config){
+    var parts = config.path.split('.');
+    var head = _.initial(parts, parts.length - 1);
+    var tail = _.last(config.path.split('.'));
```

Comment:
Why not just use `parts`?
Also I think
`_.initial(parts, parts.length - 1)` can be replaced with`_.first(parts)`